# Geocoding using googlemaps (directions)


For this exercise we got our own google api and token. they have been DEACTIVATED to evade the unapproved use of them.

In [2]:
#pip install googlemaps

In [3]:
#pip install camelot-py[cv] tabula-py

In [19]:
# Packages
import pandas as pd
import os
import urllib.request, json, csv
import numpy as np
from tqdm import tqdm_notebook as tqdm
# For sending GET requests from the API
import requests
# For saving access tokens and for file management when creating and adding to the dataset
import os
# For dealing with json responses we receive from the API
import json
# For displaying the data after
import pandas as pd
# For saving the response data in CSV format
import csv
# For parsing the dates received from twitter in readable formats
import datetime
import dateutil.parser
import unicodedata
#To add wait time between requests
import time
import requests
#maps
import googlemaps
#pdf
from tqdm import tqdm_notebook as tqdm
import tabula

In [5]:
os.environ["JAVA_HOME"] = "/path/to/java"

## 1. First we import the pdf file into pandas and only keep desired rows

Import Data from this url. This dataset is in PDF format. You have to convert to PandasDataFrame and keep only the BBVA offices in LIMA. 

We had to install Jave from this [link](https://www.java.com/en/download/) in order to be able to run **tabula**.

In [6]:
dbanco = tabula.read_pdf("https://www.bbva.pe/content/dam/public-web/peru/documents/personas/canales-de-atencion/oficinas/Oficinas-BBVA-abiertas-23.10.20.pdf", pages="all")

In [7]:
#Delete the empty  elements of this list
dbanco[1].dropna(how='all', axis=1, inplace=True)
dbanco

[                       Unnamed: 0  \
 0                         OFICINA   
 1                    BAGUA GRANDE   
 2                        CHIMBOTE   
 3                          HUARAZ   
 4              MEGAPLAZA CHIMBOTE   
 5                             NaN   
 6         MERCADO MODELO CHIMBOTE   
 7                  NUEVO CHIMBOTE   
 8                             NaN   
 9                     ANDAHUAYLAS   
 10                     LAS BAMBAS   
 11                       AREQUIPA   
 12              CAMANA - AREQUIPA   
 13                          CAYMA   
 14                 CERRO COLORADO   
 15                    EL PEDREGAL   
 16                            NaN   
 17  JOSE LUIS BUSTAMANTE Y RIVERO   
 18                            NaN   
 19       MALL AVENTURA PLAZA  AQP   
 20                            NaN   
 21          MIRAFLORES - AREQUIPA   
 22              PARQUE INDUSTRIAL   
 23                     SAN CAMILO   
 24                      YANAHUARA   
 25         

In [8]:
#order the observations in groups: "oficina", "direccion", "distrito", "provincia", "departamento
for i in range(len(dbanco)):
    dbanco[i] = dbanco[i].reset_index().T.reset_index().T.reset_index()
    dbanco[i].drop(dbanco[i].iloc[:, 0:2], axis = 1, inplace=True)
    dbanco[i].set_axis(["oficina", "direccion", "distrito", "provincia", "departamento"], axis=1, inplace=True)

In [9]:
# we eliminate the first 2 columns that were empty and wronged processed in the first page
dbanco[0]= dbanco[0].iloc[2:, :]

In [10]:
#Concatenate the data
bbva_peru = pd.concat(dbanco, ignore_index= True)
bbva_peru.head(2)

oficina                               direccion      distrito  \
0  BAGUA GRANDE                    AV. CHACHAPOYAS 2044  BAGUA GRANDE   
1      CHIMBOTE  AV. FRANCISCO BOLOGNESI 696 - CHIMBOTE      CHIMBOTE   

   provincia departamento  
0  UTCUBAMBA     AMAZONAS  
1      SANTA       ANCASH

In [11]:
#we keep only LIMA observations
bbva_peru["departamento"].replace(["LIMA.1"], ["LIMA"], inplace =   True)
print(bbva_peru['departamento'].value_counts())
df = bbva_peru.loc[bbva_peru["departamento"] == "LIMA", :]

LIMA             206
AREQUIPA          12
LA LIBERTAD       12
PIURA             10
JUNIN              9
LAMBAYEQUE         7
SAN MARTIN         6
CUSCO              6
ICA                5
ANCASH             5
LORETO             4
CAJAMARCA          4
PUNO               3
UCAYALI            3
HUANUCO            3
MOQUEGUA           2
APURIMAC           2
TACNA              2
MADRE DE DIOS      1
PASCO.1            1
AYACUCHO           1
TUMBES             1
AMAZONAS           1
Name: departamento, dtype: int64


In [13]:
#concatenate the district and the direction of the bank to help the searcher
cols = ['direccion', 'distrito']
df[ 'complete_address' ] = df[ cols ].apply(lambda row: ', '.\
                                  join( row.values.astype( str ) ) , axis = 1 )

C:\Users\DELL\AppData\Local\Temp\ipykernel_10772\3454748180.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[ 'complete_address' ] = df[ cols ].apply(lambda row: ', '.\


In [14]:
df['direc'] = df['complete_address'] + " " + "BBVA"
df.head(5)
#By adding strings from "DISTRITO", there is more information in the adress and we reduce the NANs from 39 to 7.
#Then, by adding "BBVA" to the string we reduce the missing values from 6 to 3.

C:\Users\DELL\AppData\Local\Temp\ipykernel_10772\4274332164.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['direc'] = df['complete_address'] + " " + "BBVA"


oficina                                          direccion  \
78       CAMINO REAL                                AV. CAMINO REAL 355   
79       28 DE JULIO                                AV. 28 DE JULIO 775   
80  ABANCAY-CONGRESO                                AV. ABANCAY 260-262   
81        AEROPUERTO       CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A   
82      ANGAMOS ESTE  Av. Angamos Este 215 - 227, esquina con Av. Pe...   

      distrito provincia departamento  \
78  SAN ISIDRO      LIMA         LIMA   
79        LIMA      LIMA         LIMA   
80        LIMA      LIMA         LIMA   
81      CALLAO      LIMA         LIMA   
82  MIRAFLORES      LIMA         LIMA   

                                     complete_address  \
78                    AV. CAMINO REAL 355, SAN ISIDRO   
79                          AV. 28 DE JULIO 775, LIMA   
80                          AV. ABANCAY 260-262, LIMA   
81  CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A, ...   
82  Av. Angamos Este 215 - 227, esquina con Av. Pe...   

                                                direc  
78               AV. CAMINO REAL 355, SAN ISIDRO BBVA  
79                     AV. 28 DE JULIO 775, LIMA BBVA  
80                     AV. ABANCAY 260-262, LIMA BBVA  
81  CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A, ...  
82  Av. Angamos Este 215 - 227, esquina con Av. Pe...

## 2. Then we find the coordinates for each BBVA office

Use GoogleMaps API and geocode all the BBVA offices. For those offices that Google API gets no information, use internet and get the latitude and longitude handly and add them to dataset.


In [17]:
coord = np.zeros(shape=( len(df.direc), 2), dtype =float)
i=0

In [18]:
gmaps = googlemaps.Client(key='AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g')

# Loop over lists
for d in tqdm(df.direc):   

    # Geocoding an address
    geocode_result = gmaps.geocode( d , region = 'pe')
    
    # Check the len of result
    if len(geocode_result)==0 :
        coord[i][0] = np.nan
        coord[i][1] = np.nan

    
    # Get info
    else :
        coord[i][0] = geocode_result[0]['geometry']['location']['lat']
        coord[i][1] = geocode_result[0]['geometry']['location']['lng']
        
    i=i+1

C:\Users\DELL\AppData\Local\Temp\ipykernel_10772\1261474639.py:4: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for d in tqdm(df.direc):


  0%|          | 0/206 [00:00<?, ?it/s]

In [25]:
#Create the dataframe with the new coordenates
coords = pd.DataFrame( coord.tolist() , columns = ["Lat", "Lon"] )
coords['direc'] = df.direc
coords

Lat        Lon                                              direc
0   -12.097362 -77.036099                                                NaN
1   -12.064030 -77.037961                                                NaN
2   -12.047955 -77.026997                                                NaN
3   -12.023538 -77.104397                                                NaN
4   -12.113476 -77.028582                                                NaN
..         ...        ...                                                ...
201 -12.062448 -77.093581  AV. AVIACIÓN 3328, MZ. B1 LT. 6, URB. LAS MAGN...
202 -12.062026 -77.092444                AV. IGNACIO MERINO 2099, LINCE BBVA
203 -12.038676 -76.976219  AV. PRIMAVERA 2180-2190, ESQ. CON AV. CENTRAL ...
204 -11.994069 -77.063041                                                NaN
205 -11.963811 -77.063462            ALFREDO MENDIOLA 3685., LOS OLIVOS BBVA

[206 rows x 3 columns]

In [26]:
#Now we merge the coordenates dataframe with the original dataframe into the final df
df_final = df.merge(coords, left_on='direc', right_on='direc', how='left')
df_final.drop('direc', inplace=True, axis=1)
df_final.index.name = 'id'
df_final.reset_index(inplace=True)
df_final

id                 oficina  \
0      0             CAMINO REAL   
1      1             28 DE JULIO   
2      2        ABANCAY-CONGRESO   
3      3              AEROPUERTO   
4      4            ANGAMOS ESTE   
..   ...                     ...   
201  201     NESTLE-DONOFRIO 164   
202  202  BIENESTAR DE LA MARINA   
203  203                 SEDAPAL   
204  204         ESPECIAL SENATI   
205  205          VILLA INFANTAS   

                                             direccion       distrito  \
0                                  AV. CAMINO REAL 355     SAN ISIDRO   
1                                  AV. 28 DE JULIO 775           LIMA   
2                                  AV. ABANCAY 260-262           LIMA   
3         CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A         CALLAO   
4    Av. Angamos Este 215 - 227, esquina con Av. Pe...     MIRAFLORES   
..                                                 ...            ...   
201                                 AV. VENEZUELA 2580           LIMA   
202                          AV. VENEZUELA S/N CDRA 34     SAN MIGUEL   
203  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...    EL AGUSTINO   
204                          AV. ALFREDO MENDIOLA 3520  INDEPENDENCIA   
205  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...     LOS OLIVOS   

    provincia departamento                                   complete_address  \
0        LIMA         LIMA                    AV. CAMINO REAL 355, SAN ISIDRO   
1        LIMA         LIMA                          AV. 28 DE JULIO 775, LIMA   
2        LIMA         LIMA                          AV. ABANCAY 260-262, LIMA   
3        LIMA         LIMA  CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A, ...   
4        LIMA         LIMA  Av. Angamos Este 215 - 227, esquina con Av. Pe...   
..        ...          ...                                                ...   
201      LIMA         LIMA                           AV. VENEZUELA 2580, LIMA   
202      LIMA         LIMA              AV. VENEZUELA S/N CDRA 34, SAN MIGUEL   
203      LIMA         LIMA  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...   
204      LIMA         LIMA           AV. ALFREDO MENDIOLA 3520, INDEPENDENCIA   
205      LIMA         LIMA  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...   

           Lat        Lon  
0   -12.067611 -77.115078  
1   -12.055685 -77.096612  
2   -11.107431 -77.608089  
3   -11.496517 -77.211773  
4   -12.016421 -76.825004  
..         ...        ...  
201        NaN        NaN  
202        NaN        NaN  
203        NaN        NaN  
204        NaN        NaN  
205        NaN        NaN  

[206 rows x 9 columns]

### Now we replace NAs by hand

In [27]:
isnan = df_final[df_final.isna().any(axis=1)]
len(isnan)
#There are 3 cases with missing values: Mala, Naciones Unidas & Strip Center Barranca

101

In [28]:
isnan

id                 oficina  \
33    33                 CAMACHO   
45    45      CC. REAL PLAZA PRO   
79    79         HIPOLITO UNANUE   
92    92               LA MARINA   
109  109              LOS ROBLES   
..   ...                     ...   
201  201     NESTLE-DONOFRIO 164   
202  202  BIENESTAR DE LA MARINA   
203  203                 SEDAPAL   
204  204         ESPECIAL SENATI   
205  205          VILLA INFANTAS   

                                             direccion              distrito  \
33                    AV. PROL. JAVIER PRADO ESTE 4921             LA MOLINA   
45             AV. ALFREDO MENDIOLA 7038-7042, LC B-02  SAN MARTIN DE PORRES   
79                  JR HIPOLITO UNANUE 1616-1620, LC 3           LA VICTORIA   
92   AV. LA MARINA N° 2355 CENTRO COMERCIAL LA MARI...            SAN MIGUEL   
109                               AV. DOS DE MAYO 1198            SAN ISIDRO   
..                                                 ...                   ...   
201                                 AV. VENEZUELA 2580                  LIMA   
202                          AV. VENEZUELA S/N CDRA 34            SAN MIGUEL   
203  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...           EL AGUSTINO   
204                          AV. ALFREDO MENDIOLA 3520         INDEPENDENCIA   
205  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...            LOS OLIVOS   

    provincia departamento                                   complete_address  \
33       LIMA         LIMA        AV. PROL. JAVIER PRADO ESTE 4921, LA MOLINA   
45       LIMA         LIMA  AV. ALFREDO MENDIOLA 7038-7042, LC B-02, SAN M...   
79       LIMA         LIMA    JR HIPOLITO UNANUE 1616-1620, LC 3, LA VICTORIA   
92       LIMA         LIMA  AV. LA MARINA N° 2355 CENTRO COMERCIAL LA MARI...   
109      LIMA         LIMA                   AV. DOS DE MAYO 1198, SAN ISIDRO   
..        ...          ...                                                ...   
201      LIMA         LIMA                           AV. VENEZUELA 2580, LIMA   
202      LIMA         LIMA              AV. VENEZUELA S/N CDRA 34, SAN MIGUEL   
203      LIMA         LIMA  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...   
204      LIMA         LIMA           AV. ALFREDO MENDIOLA 3520, INDEPENDENCIA   
205      LIMA         LIMA  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...   

     Lat  Lon  
33   NaN  NaN  
45   NaN  NaN  
79   NaN  NaN  
92   NaN  NaN  
109  NaN  NaN  
..   ...  ...  
201  NaN  NaN  
202  NaN  NaN  
203  NaN  NaN  
204  NaN  NaN  
205  NaN  NaN  

[101 rows x 9 columns]

In [29]:
#Oficina Mala
df_final.loc[113,'Lat']= -12.6558538
df_final.loc[113,'Lon']= -76.6348772
#Oficina Naciones Unidas
df_final.loc[127,'Lat']= -12.1003510
df_final.loc[127,'Lon']= -77.0636233
#Oficina Strip Center Barranca
df_final.loc[40,'Lat']= -10.75237060
df_final.loc[40,'Lon']= -77.75986317

## 3. Find driving time

Use Google API to find the driving time (best guess) from all the group members' address and all the LIMA BBVA offices.

Les puse direcciones al azar. Sería cuestión que lo editen.

#### Group members' address
- Alejandro: -12.1148761,-77.0386600
- Steph: -12.1148389,-77.0117565
- Juan Carlos: -12.0961813,-77.0714548
- Gabriela: -12.0259368,-77.0069487

In [34]:
#We subset only coordenates and transform them to strings instead of floats
bbva = df_final.iloc[:, [7,8]].applymap(str)
bbva.head(2)

Lat          Lon
0  -12.0676108  -77.1150782
1  -12.0556849  -77.0966119

In [35]:
#Create a list of "destinations"
bbva['destino'] = bbva['Lat'] + "," + bbva['Lon']
destino = bbva['destino'].tolist()
destino

['-12.0676108,-77.1150782',
 '-12.0556849,-77.0966119',
 '-11.1074314,-77.6080894',
 '-11.4965172,-77.21177329999999',
 '-12.016421,-76.8250038',
 '-12.0279109,-77.0575418',
 '-11.9916265,-77.07262639999999',
 '-12.0317332,-77.0712543',
 '-12.0738899,-77.0489924',
 '-12.0851775,-76.97939810000001',
 '-12.094951,-77.03747',
 '-12.0496302,-77.0375281',
 '-12.1012422,-76.9706199',
 '-12.1111508,-76.9873252',
 '-12.0775066,-77.0870468',
 '-12.0678936,-76.9584569',
 '-12.0598246,-77.1351292',
 '-12.077758,-76.91851199999999',
 '-13.5295906,-71.96458539999999',
 '-12.0737038,-76.9560571',
 '-12.0619569,-77.02961669999999',
 '-12.1235885,-77.02911619999999',
 '-12.0922232,-77.0236816',
 '-12.007389,-77.0054625',
 '-12.0518113,-77.03360909999999',
 '-12.0379707,-77.0986351',
 '-12.1037833,-77.001634',
 '-12.0852949,-77.03225119999999',
 '-12.1053898,-76.9654438',
 '-11.976987,-77.067587',
 '-12.0917418,-77.04293109999999',
 '-12.0982554,-76.9625544',
 '-12.0922212,-77.0716281',
 'nan,nan',
 '-

In [36]:
#Alejandro
df_driving = pd.DataFrame (destino, columns = ['destination'])
df_driving['origin'] = "-12.1148761,-77.0386600"
df_driving

destination                   origin
0           -12.0676108,-77.1150782  -12.1148761,-77.0386600
1           -12.0556849,-77.0966119  -12.1148761,-77.0386600
2           -11.1074314,-77.6080894  -12.1148761,-77.0386600
3    -11.4965172,-77.21177329999999  -12.1148761,-77.0386600
4            -12.016421,-76.8250038  -12.1148761,-77.0386600
..                              ...                      ...
201                         nan,nan  -12.1148761,-77.0386600
202                         nan,nan  -12.1148761,-77.0386600
203                         nan,nan  -12.1148761,-77.0386600
204                         nan,nan  -12.1148761,-77.0386600
205                         nan,nan  -12.1148761,-77.0386600

[206 rows x 2 columns]

In [37]:
#Steph
df_driving2 = pd.DataFrame (destino, columns = ['destination'])
df_driving2['origin'] = "-12.1148389,-77.0117565"
df_driving2

destination                   origin
0           -12.0676108,-77.1150782  -12.1148389,-77.0117565
1           -12.0556849,-77.0966119  -12.1148389,-77.0117565
2           -11.1074314,-77.6080894  -12.1148389,-77.0117565
3    -11.4965172,-77.21177329999999  -12.1148389,-77.0117565
4            -12.016421,-76.8250038  -12.1148389,-77.0117565
..                              ...                      ...
201                         nan,nan  -12.1148389,-77.0117565
202                         nan,nan  -12.1148389,-77.0117565
203                         nan,nan  -12.1148389,-77.0117565
204                         nan,nan  -12.1148389,-77.0117565
205                         nan,nan  -12.1148389,-77.0117565

[206 rows x 2 columns]

In [38]:
#Gabriela
df_driving3 = pd.DataFrame (destino, columns = ['destination'])
df_driving3['origin'] = "-12.0259368,-77.0069487"
df_driving3

destination                   origin
0           -12.0676108,-77.1150782  -12.0259368,-77.0069487
1           -12.0556849,-77.0966119  -12.0259368,-77.0069487
2           -11.1074314,-77.6080894  -12.0259368,-77.0069487
3    -11.4965172,-77.21177329999999  -12.0259368,-77.0069487
4            -12.016421,-76.8250038  -12.0259368,-77.0069487
..                              ...                      ...
201                         nan,nan  -12.0259368,-77.0069487
202                         nan,nan  -12.0259368,-77.0069487
203                         nan,nan  -12.0259368,-77.0069487
204                         nan,nan  -12.0259368,-77.0069487
205                         nan,nan  -12.0259368,-77.0069487

[206 rows x 2 columns]

In [39]:
#Juan Carlos
df_driving4 = pd.DataFrame (destino, columns = ['destination'])
df_driving4['origin'] = "-12.0961813,-77.0714548"
df_driving4

destination                   origin
0           -12.0676108,-77.1150782  -12.0961813,-77.0714548
1           -12.0556849,-77.0966119  -12.0961813,-77.0714548
2           -11.1074314,-77.6080894  -12.0961813,-77.0714548
3    -11.4965172,-77.21177329999999  -12.0961813,-77.0714548
4            -12.016421,-76.8250038  -12.0961813,-77.0714548
..                              ...                      ...
201                         nan,nan  -12.0961813,-77.0714548
202                         nan,nan  -12.0961813,-77.0714548
203                         nan,nan  -12.0961813,-77.0714548
204                         nan,nan  -12.0961813,-77.0714548
205                         nan,nan  -12.0961813,-77.0714548

[206 rows x 2 columns]

In [40]:
df_driving_final = pd.concat([df_driving, df_driving2, df_driving3, df_driving4]).reset_index(drop=True)
df_driving_final

destination                   origin
0           -12.0676108,-77.1150782  -12.1148761,-77.0386600
1           -12.0556849,-77.0966119  -12.1148761,-77.0386600
2           -11.1074314,-77.6080894  -12.1148761,-77.0386600
3    -11.4965172,-77.21177329999999  -12.1148761,-77.0386600
4            -12.016421,-76.8250038  -12.1148761,-77.0386600
..                              ...                      ...
819                         nan,nan  -12.0961813,-77.0714548
820                         nan,nan  -12.0961813,-77.0714548
821                         nan,nan  -12.0961813,-77.0714548
822                         nan,nan  -12.0961813,-77.0714548
823                         nan,nan  -12.0961813,-77.0714548

[824 rows x 2 columns]

In [41]:
df_driving_final['combination'] = list(range(1,len(df_driving_final)+1))
df_driving_final

destination                   origin  combination
0           -12.0676108,-77.1150782  -12.1148761,-77.0386600            1
1           -12.0556849,-77.0966119  -12.1148761,-77.0386600            2
2           -11.1074314,-77.6080894  -12.1148761,-77.0386600            3
3    -11.4965172,-77.21177329999999  -12.1148761,-77.0386600            4
4            -12.016421,-76.8250038  -12.1148761,-77.0386600            5
..                              ...                      ...          ...
819                         nan,nan  -12.0961813,-77.0714548          820
820                         nan,nan  -12.0961813,-77.0714548          821
821                         nan,nan  -12.0961813,-77.0714548          822
822                         nan,nan  -12.0961813,-77.0714548          823
823                         nan,nan  -12.0961813,-77.0714548          824

[824 rows x 3 columns]

In [42]:
#Create the three lists: combination, origin and destination
combination = df_driving_final['combination'].tolist()
origin = df_driving_final['origin'].tolist()
destination = df_driving_final['destination'].tolist()

In [43]:
data_distance = {} 

In [44]:
# Loop to generate info about geolocations
duration_info = np.zeros(shape=(len(combination),2), dtype =float)
i=0

for c,o,d in tqdm(list(zip(combination, origin, destination))):
    try:
        # Google MapsDdirections API endpoint
        endpoint = 'https://maps.googleapis.com/maps/api/directions/json?'

    ## Fixed Parameters
        # Paramaters
        traffic_model = 'best_guess'  

        # Departure time
        departure_time = 'now'

        # driving, walking, biclycling, transit
        mode = 'driving'

        # key
        api_key = 'AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g'

        # region to look for (spain= es , germany = de , Switzerland= swiss)
        region = 'pe'

     ## Parameters
        # Origin
        origin = o
        
        # Destinations
        destination = d
        
        #Building the URL for the request
        nav_request = 'origin={}&destination={}&departure_time={}&traffic_model={}&mode={}&region={}&key={}'.format(origin , 
                        destination , departure_time , traffic_model , mode, region, api_key)
        
        # Concatenate strings
        request = endpoint + nav_request
        print(request)

        #Sends the request and reads the response.
        response = urllib.request.urlopen(request).read()
        
        

        #Loads response as JSON
        directions = json.loads(response)

        legs = directions['routes'][0]['legs'][0]
        #print(legs)
        
        duration_info[i][0] = c
        duration_info[i][1] = legs['duration']['value']
        #print(duration_info[i][1])
        
        i=i+1

        my_keys = ['duration']
        info = { my_key: legs[my_key] for my_key in my_keys }
        data_distance[c] = info 
    
    except Exception as e:
        
        duration_info[i][0] = c
        duration_info[i][1] = "nan"
        
        i=i+1
        
        print(c)

C:\Users\DELL\AppData\Local\Temp\ipykernel_10772\680288981.py:5: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for c,o,d in tqdm(list(zip(combination, origin, destination))):


  0%|          | 0/824 [00:00<?, ?it/s]

https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0676108,-77.1150782&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0556849,-77.0966119&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-11.1074314,-77.6080894&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-11.4965172,-77.21177329999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.11487

https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.1512936,-76.9817363&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0779212,-77.0924843&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0433031,-76.97036729999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-11.9933447,-77.0630175&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.11487

https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0762692,-76.9872565&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0995207,-77.0564507&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.025525,-77.03315549999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0897789,-77.0030872&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.114876

https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-11.9638112,-77.063462&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
110
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
111
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
112
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure

146
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
147
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
148
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
149
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
150
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=n

184
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
185
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
186
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
187
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
188
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=nan,nan&departure_time=n

https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-12.0678936,-76.9584569&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-12.0598246,-77.1351292&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-12.077758,-76.91851199999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-13.5295906,-71.96458539999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12

https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-12.0218001,-77.0596169&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-12.0776021,-77.0863292&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-12.1192591,-77.03667899999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-12.0736141,-77.0164831&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.11483

https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-12.1073943,-77.01579079999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-11.9988068,-77.0626908&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-12.0814037,-77.08250559999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-11.944697,-77.028683&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.

332
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
333
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=-12.100351,-77.0636233&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
335
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
336
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&depar

370
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
371
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
372
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
373
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
374
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=n

408
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
409
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
410
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
411
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148389,-77.0117565&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
412
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-12.0676108,-77.1150782&

https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
446
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-12.0565632,-77.1015012&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-12.0566677,-76.97018369999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-12.1512936,-76.9817363&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.00694

https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-12.1192904,-77.0268905&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-12.0823232,-76.92840249999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-12.0762692,-76.9872565&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-12.0995207,-77.0564507&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.02593

https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-12.0620255,-77.0924442&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-12.0386756,-76.976219&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=-11.9638112,-77.063462&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
522
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destin

556
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
557
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
558
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
559
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
560
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=n

594
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
595
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
596
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
597
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
598
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0259368,-77.0069487&destination=nan,nan&departure_time=n

https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.0775066,-77.0870468&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.0678936,-76.9584569&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.0598246,-77.1351292&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.077758,-76.91851199999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.096181

https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.1010005,-77.01074489999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.0218001,-77.0596169&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.0776021,-77.0863292&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.1192591,-77.03667899999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-1

https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.1073943,-77.01579079999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-11.9988068,-77.0626908&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.0814037,-77.08250559999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-11.944697,-77.028683&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.

744
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
745
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=-12.100351,-77.0636233&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
747
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
748
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&depar

782
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
783
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
784
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
785
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
786
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=n

820
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
821
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
822
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
823
https://maps.googleapis.com/maps/api/directions/json?origin=-12.0961813,-77.0714548&destination=nan,nan&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
824


In [ ]:
duration_info

In [ ]:
distance_api = pd.DataFrame(duration_info, columns = ['Combination', 'Duration_seg'])
distance_api

## 4. Report offices closets and furthest from each group members' adress

Se me ocurre que podríamos crear un dataframe en base a lo anterior (cuando se arregle) que sea una columna por miembro del grupo con el tiempo a cada una de las 205 oficinas, y hacemos un merge con la df_final. Luego que nos dé la fila donde está el valor mínimo y máximo para cada una de las 4 columnas y así identificamos las oficinas.

In [ ]:
group_size = 206  # Set group size to 206
num_groups = (len(distance_api) - 1) // group_size + 1  # Calculate number of groups needed
distance_api['group'] = [i // group_size + 1 for i in range(len(distance_api))]  # Assign group numbers


In [ ]:
distance_api

In [ ]:
#Alejandro
distance_api_1 = distance_api[distance_api["group"] == 1]
distance_api_1.sort_values(by=['Duration_seg'])

The BBVAs closest to Alejandro's house are those belonging to the combination with the 175, 139 and 56 BBVA in the list. While the farthest are those belonging to the combination with the 149, 93 and 159 BBVA in the list. The nearest is the Plaza Union office, and the fartest is the Santa Monica office. 

In [ ]:
df_list = df[(df.index == 175) | (df.index == 149)]
df_list

In [ ]:
#Steph
distance_api_2 = distance_api[distance_api["group"] == 2]
distance_api_2.sort_values(by=['Duration_seg'])

The BBVAs closest to Stephy's house are those belonging to the combination with the 24, 7 and 21 BBVA in the list (229, 212 and 226). While the farthest are those belonging to the combination with the 150, 94 and 160 BBVA in the list (355, 299 and 365). The nearest is the C.C Fiori office and the fartest is the Parque Industrial in Villa El Salvador office. 

In [ ]:
df_list_2 = df[(df.index == 24) | (df.index == 150)]
df_list_2

In [ ]:
#Gabriela
distance_api_2 = distance_api[distance_api["group"] == 3]
distance_api_2.sort_values(by=['Duration_seg'])

The BBVAs closest to Gabriela's house are those belonging to the combination with the 130, 116 and 88  BBVAs of the lista (540, 526 and 498), while the farthest are those belonging to the combination with 151, 95 and 191 BBVA of the list (561, 505 and 571).The nearest is the Nicolas Ayllon office and the fartest is the Pro office.

In [ ]:
df_list_3 = df[(df.index == 130) | (df.index == 151)]
df_list_3

In [ ]:
#JC
distance_api_4 = distance_api[distance_api["group"] == 4]
distance_api_4.sort_values(by=['Duration_seg'])

The BBVAs closest to Juan Carlos' house are those belonging to the combination with the 131, 116 and 89 BBVAs of the list (746, 731 and 704). While the farthest are those belonging to the combination with the 162, 96 and 152 BBVA of the list (777, 711 and 767). The nearest is the Orrantia office and the fartest is the Samuel del Alcalzar office.

In [ ]:
df_list_4 = df[(df.index == 131) | (df.index == 162)]
df_list_4